interfaces/builtin/pwm_test.go

```diff
@@ -0,0 +1,207 @@
+slots:
+    my-pin:
+        interface: pwm
+        chip-number: 10
+        number: 7
+`, nil)
+	s.osPwmSlotInfo = osInfo.Slots["my-pin"]
+	s.osPwmSlot = interfaces.NewConnectedSlot(s.osPwmSlotInfo, nil, nil)
+}
+
+func (s *PwmInterfaceSuite) TestName(c *C) {
+	c.Assert(s.iface.Name(), Equals, "pwm")
+}
+
+func (s *PwmInterfaceSuite) TestSanitizeSlotGadgetSnap(c *C) {
+	// pwm slot on gadget accepeted
+	c.Assert(interfaces.BeforePrepareSlot(s.iface, s.gadgetPwmSlotInfo), IsNil)
+
+	// slots without number attribute are rejected
+	c.Assert(interfaces.BeforePrepareSlot(s.iface, s.gadgetMissingNumberSlotInfo), ErrorMatches,
+		"pwm slot must have a number attribute")
+
+	// slots with number attribute that isnt a number
+	c.Assert(interfaces.BeforePrepareSlot(s.iface, s.gadgetBadNumberSlotInfo), ErrorMatches,
+		"pwm slot number attribute must be an int")
+
+	// slots without number attribute are rejected
+	c.Assert(interfaces.BeforePrepareSlot(s.iface, s.gadgetMissingChipNumberSlotInfo), ErrorMatches,
+		"pwm slot must have a chip-number attribute")
+
+	// slots with number attribute that isnt a number
+	c.Assert(interfaces.BeforePrepareSlot(s.iface, s.gadgetBadChipNumberSlotInfo), ErrorMatches,
+		"pwm slot chip-number attribute must be an int")
+}
+
+func (s *PwmInterfaceSuite) TestSanitizeSlotOsSnap(c *C) {
+	// pwm slot on OS accepted
+	c.Assert(interfaces.BeforePrepareSlot(s.iface, s.osPwmSlotInfo), IsNil)
+}
+
+func (s *PwmInterfaceSuite) TestSanitizePlug(c *C) {
+	c.Assert(interfaces.BeforePreparePlug(s.iface, s.gadgetPlugInfo), IsNil)
+}
+
+func (s *PwmInterfaceSuite) TestSystemdConnectedSlot(c *C) {
+	spec := &systemd.Specification{}
+	err := spec.AddConnectedSlot(s.iface, s.gadgetPlug, s.gadgetPwmSlot)
+	c.Assert(err, IsNil)
+	c.Assert(spec.Services(), DeepEquals, map[string]*systemd.Service{
+		"snap.my-device.interface.pwmchip10-pwm100.service": {
+			Type:            "oneshot",
+			RemainAfterExit: true,
+			ExecStart:       `/bin/sh -c 'test -e /sys/class/pwm/pwmchip10/pwm100 || echo 100 > /sys/class/pwm/pwmchip10/export'`,
+			ExecStop:        `/bin/sh -c 'test ! -e /sys/class/pwm/pwmchip10/pwm100 || echo 100 > /sys/class/pwm/pwmchip10/unexport'`,
+		},
+	})
+}
+
+func (s *PwmInterfaceSuite) TestApparmorConnectedPlugIgnoresMissingSymlink(c *C) {
+	log, restore := logger.MockLogger()
+	defer restore()
+
+	builtin.MockEvalSymlinks(&s.BaseTest, func(path string) (string, error) {
+		c.Assert(path, Equals, "/sys/class/pwm/pwmchip10")
+		return "", os.ErrNotExist
+	})
+
+	spec := &apparmor.Specification{}
+	err := spec.AddConnectedPlug(s.iface, s.gadgetPlug, s.gadgetPwmSlot)
+	c.Assert(err, IsNil)
+	c.Assert(spec.Snippets(), HasLen, 0)
+	c.Assert(log.String(), testutil.Contains, "cannot find not existing pwm chipbase /sys/class/pwm/pwmchip10")
+}
+
+func (s *PwmInterfaceSuite) TestInterfaces(c *C) {
+	c.Check(builtin.Interfaces(), testutil.DeepContains, s.iface)
+}
+
+func (s *PwmInterfaceSuite) TestApparmorConnectedSlot(c *C) {
```

Comment:
as was pointed out in my PR, this should actually be called:

```suggestion
func (s *PwmInterfaceSuite) TestApparmorConnectedPlug(c *C) {
```

but don't worry about updating it here, we can do it in a followup or I can push a commit here if tests are red due to unrelated failures and would need to be retriggered anyways

Reply:
Let's get it updated!